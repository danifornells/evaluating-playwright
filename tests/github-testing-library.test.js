/**
 * Provided example by JEST-PLAYWRIGHT,
 * rewritten as TESTING-LIBRARY recommendations
 * https://github.com/playwright-community/playwright-jest-examples/blob/master/basic/tests/github.test.js
 *
 * What a cocktail 🍹
 *
 * JEST ......................... https://jestjs.io/docs/en/api
 * JEST-DOM ..................... https://github.com/testing-library/jest-dom
 * PLAYWRIGHT ................... https://playwright.dev/#?path=docs/api.md
 * JEST-PLAYWRIGHT .............. https://github.com/playwright-community/jest-playwright
 * |- EXPECT-PLAYWRIGHT ......... https://github.com/playwright-community/expect-playwright
 * TESTING-LIBRARY .............. https://testing-library.com/docs/intro
 * PLAYWRIGHT-TESTING-LIBRARY ... https://github.com/hoverinc/playwright-testing-library
 *
 * **/

const {getDocument, queries} = require('playwright-testing-library')
const {findByPlaceholderText, findByRole} = queries

describe("GitHub", ()=> {
  it("should show the microsoft/Playwright project in the search if you search for Playwright", async()=> {
    await page.goto("https://github.com");
    const $homePage = await getDocument(page);
    const $searchField = await findByPlaceholderText($homePage, 'search',{ exact: false });
    await $searchField.type("Playwright");
    await $searchField.press("Enter");

    // Following statements fails:
    //
    // const $resultsDocument = await getDocument(page); Needed?
    // await findByText($resultsDocument, "microsoft/playwright", { exact: false });
    //
    // Evaluation failed: Error: Unable to find an element with the text: microsoft/playwright.
    // This could be because the text is broken up by multiple elements.
    // In this case, you can provide a function for your text matcher to make your matcher more flexible.
    //
    // https://github.com/testing-library/dom-testing-library/issues/410
    // Solved by using findByRole

    // Solved by using Playwright
    // await expect(page).toHaveText(".repo-list a", "microsoft/playwright");

    // Solved by using findByRole
    const $searchPage = await getDocument(page);
    await findByRole($searchPage, "link", { name: "microsoft/playwright" });
  })

  it("should contain 'Playwright' in the project title", async () => {
    const $searchPage = await getDocument(page);
    const $repoLink = await findByRole($searchPage, "link", { name: "microsoft/playwright" });
    await $repoLink.click();
    await expect(page).toHaveText("#readme h1", "Playwright");
  })

  it("should contain a tab nav as expected", async () => {
    const $tabNav = await page.$('.js-repo-nav')
    const tabNavHtml = await $tabNav.innerHTML()
    expect(tabNavHtml).toMatchSnapshot()
  })
})