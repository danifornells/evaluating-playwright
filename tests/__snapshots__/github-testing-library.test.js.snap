// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitHub should contain a tab nav as expected 1`] = `
"
  <ul class=\\"UnderlineNav-body list-style-none \\">
          <li class=\\"d-flex\\">
        <a class=\\"js-selected-navigation-item selected UnderlineNav-item hx_underlinenav-item no-wrap js-responsive-underlinenav-item\\" data-tab-item=\\"code-tab\\" data-hotkey=\\"g c\\" data-ga-click=\\"Repository, Navigation click, Code tab\\" aria-current=\\"page\\" data-selected-links=\\"repo_source repo_downloads repo_commits repo_releases repo_tags repo_branches repo_packages repo_deployments /microsoft/playwright\\" href=\\"/microsoft/playwright\\">
              <svg classes=\\"UnderlineNav-octicon\\" display=\\"none inline\\" height=\\"16\\" class=\\"octicon octicon-code UnderlineNav-octicon d-none d-sm-inline\\" viewBox=\\"0 0 16 16\\" version=\\"1.1\\" width=\\"16\\" aria-hidden=\\"true\\"><path fill-rule=\\"evenodd\\" d=\\"M4.72 3.22a.75.75 0 011.06 1.06L2.06 8l3.72 3.72a.75.75 0 11-1.06 1.06L.47 8.53a.75.75 0 010-1.06l4.25-4.25zm6.56 0a.75.75 0 10-1.06 1.06L13.94 8l-3.72 3.72a.75.75 0 101.06 1.06l4.25-4.25a.75.75 0 000-1.06l-4.25-4.25z\\"></path></svg>
            <span data-content=\\"Code\\">Code</span>
              <span title=\\"Not available\\" class=\\"Counter \\"></span>
</a>      </li>
      <li class=\\"d-flex\\">
        <a class=\\"js-selected-navigation-item UnderlineNav-item hx_underlinenav-item no-wrap js-responsive-underlinenav-item\\" data-tab-item=\\"issues-tab\\" data-hotkey=\\"g i\\" data-ga-click=\\"Repository, Navigation click, Issues tab\\" data-selected-links=\\"repo_issues repo_labels repo_milestones /microsoft/playwright/issues\\" href=\\"/microsoft/playwright/issues\\">
              <svg classes=\\"UnderlineNav-octicon\\" display=\\"none inline\\" height=\\"16\\" class=\\"octicon octicon-issue-opened UnderlineNav-octicon d-none d-sm-inline\\" viewBox=\\"0 0 16 16\\" version=\\"1.1\\" width=\\"16\\" aria-hidden=\\"true\\"><path fill-rule=\\"evenodd\\" d=\\"M8 1.5a6.5 6.5 0 100 13 6.5 6.5 0 000-13zM0 8a8 8 0 1116 0A8 8 0 010 8zm9 3a1 1 0 11-2 0 1 1 0 012 0zm-.25-6.25a.75.75 0 00-1.5 0v3.5a.75.75 0 001.5 0v-3.5z\\"></path></svg>
            <span data-content=\\"Issues\\">Issues</span>
              <span title=\\"149\\" class=\\"Counter \\">149</span>
</a>      </li>
      <li class=\\"d-flex\\">
        <a class=\\"js-selected-navigation-item UnderlineNav-item hx_underlinenav-item no-wrap js-responsive-underlinenav-item\\" data-tab-item=\\"pull-requests-tab\\" data-hotkey=\\"g p\\" data-ga-click=\\"Repository, Navigation click, Pull requests tab\\" data-selected-links=\\"repo_pulls checks /microsoft/playwright/pulls\\" href=\\"/microsoft/playwright/pulls\\">
              <svg classes=\\"UnderlineNav-octicon\\" display=\\"none inline\\" height=\\"16\\" class=\\"octicon octicon-git-pull-request UnderlineNav-octicon d-none d-sm-inline\\" viewBox=\\"0 0 16 16\\" version=\\"1.1\\" width=\\"16\\" aria-hidden=\\"true\\"><path fill-rule=\\"evenodd\\" d=\\"M7.177 3.073L9.573.677A.25.25 0 0110 .854v4.792a.25.25 0 01-.427.177L7.177 3.427a.25.25 0 010-.354zM3.75 2.5a.75.75 0 100 1.5.75.75 0 000-1.5zm-2.25.75a2.25 2.25 0 113 2.122v5.256a2.251 2.251 0 11-1.5 0V5.372A2.25 2.25 0 011.5 3.25zM11 2.5h-1V4h1a1 1 0 011 1v5.628a2.251 2.251 0 101.5 0V5A2.5 2.5 0 0011 2.5zm1 10.25a.75.75 0 111.5 0 .75.75 0 01-1.5 0zM3.75 12a.75.75 0 100 1.5.75.75 0 000-1.5z\\"></path></svg>
            <span data-content=\\"Pull requests\\">Pull requests</span>
              <span title=\\"18\\" class=\\"Counter \\">18</span>
</a>      </li>
      <li class=\\"d-flex\\">
        <a class=\\"js-selected-navigation-item UnderlineNav-item hx_underlinenav-item no-wrap js-responsive-underlinenav-item\\" data-tab-item=\\"actions-tab\\" data-hotkey=\\"g a\\" data-ga-click=\\"Repository, Navigation click, Actions tab\\" data-selected-links=\\"repo_actions /microsoft/playwright/actions\\" href=\\"/microsoft/playwright/actions\\">
              <svg classes=\\"UnderlineNav-octicon\\" display=\\"none inline\\" height=\\"16\\" class=\\"octicon octicon-play UnderlineNav-octicon d-none d-sm-inline\\" viewBox=\\"0 0 16 16\\" version=\\"1.1\\" width=\\"16\\" aria-hidden=\\"true\\"><path fill-rule=\\"evenodd\\" d=\\"M1.5 8a6.5 6.5 0 1113 0 6.5 6.5 0 01-13 0zM8 0a8 8 0 100 16A8 8 0 008 0zM6.379 5.227A.25.25 0 006 5.442v5.117a.25.25 0 00.379.214l4.264-2.559a.25.25 0 000-.428L6.379 5.227z\\"></path></svg>
            <span data-content=\\"Actions\\">Actions</span>
              <span title=\\"Not available\\" class=\\"Counter \\"></span>
</a>      </li>
      <li class=\\"d-flex\\">
        <a class=\\"js-selected-navigation-item UnderlineNav-item hx_underlinenav-item no-wrap js-responsive-underlinenav-item\\" data-tab-item=\\"projects-tab\\" data-hotkey=\\"g b\\" data-ga-click=\\"Repository, Navigation click, Projects tab\\" data-selected-links=\\"repo_projects new_repo_project repo_project /microsoft/playwright/projects\\" href=\\"/microsoft/playwright/projects\\">
              <svg classes=\\"UnderlineNav-octicon\\" display=\\"none inline\\" height=\\"16\\" class=\\"octicon octicon-project UnderlineNav-octicon d-none d-sm-inline\\" viewBox=\\"0 0 16 16\\" version=\\"1.1\\" width=\\"16\\" aria-hidden=\\"true\\"><path fill-rule=\\"evenodd\\" d=\\"M1.75 0A1.75 1.75 0 000 1.75v12.5C0 15.216.784 16 1.75 16h12.5A1.75 1.75 0 0016 14.25V1.75A1.75 1.75 0 0014.25 0H1.75zM1.5 1.75a.25.25 0 01.25-.25h12.5a.25.25 0 01.25.25v12.5a.25.25 0 01-.25.25H1.75a.25.25 0 01-.25-.25V1.75zM11.75 3a.75.75 0 00-.75.75v7.5a.75.75 0 001.5 0v-7.5a.75.75 0 00-.75-.75zm-8.25.75a.75.75 0 011.5 0v5.5a.75.75 0 01-1.5 0v-5.5zM8 3a.75.75 0 00-.75.75v3.5a.75.75 0 001.5 0v-3.5A.75.75 0 008 3z\\"></path></svg>
            <span data-content=\\"Projects\\">Projects</span>
              <span title=\\"0\\" hidden=\\"hidden\\" class=\\"Counter \\">0</span>
</a>      </li>
      <li class=\\"d-flex\\">
        <a class=\\"js-selected-navigation-item UnderlineNav-item hx_underlinenav-item no-wrap js-responsive-underlinenav-item\\" data-tab-item=\\"security-tab\\" data-hotkey=\\"g s\\" data-ga-click=\\"Repository, Navigation click, Security tab\\" data-selected-links=\\"security overview alerts policy token_scanning code_scanning /microsoft/playwright/security\\" href=\\"/microsoft/playwright/security\\">
              <svg classes=\\"UnderlineNav-octicon\\" display=\\"none inline\\" height=\\"16\\" class=\\"octicon octicon-shield UnderlineNav-octicon d-none d-sm-inline\\" viewBox=\\"0 0 16 16\\" version=\\"1.1\\" width=\\"16\\" aria-hidden=\\"true\\"><path fill-rule=\\"evenodd\\" d=\\"M7.467.133a1.75 1.75 0 011.066 0l5.25 1.68A1.75 1.75 0 0115 3.48V7c0 1.566-.32 3.182-1.303 4.682-.983 1.498-2.585 2.813-5.032 3.855a1.7 1.7 0 01-1.33 0c-2.447-1.042-4.049-2.357-5.032-3.855C1.32 10.182 1 8.566 1 7V3.48a1.75 1.75 0 011.217-1.667l5.25-1.68zm.61 1.429a.25.25 0 00-.153 0l-5.25 1.68a.25.25 0 00-.174.238V7c0 1.358.275 2.666 1.057 3.86.784 1.194 2.121 2.34 4.366 3.297a.2.2 0 00.154 0c2.245-.956 3.582-2.104 4.366-3.298C13.225 9.666 13.5 8.36 13.5 7V3.48a.25.25 0 00-.174-.237l-5.25-1.68zM9 10.5a1 1 0 11-2 0 1 1 0 012 0zm-.25-5.75a.75.75 0 10-1.5 0v3a.75.75 0 001.5 0v-3z\\"></path></svg>
            <span data-content=\\"Security\\">Security</span>
              
</a>      </li>
      <li class=\\"d-flex\\">
        <a class=\\"js-selected-navigation-item UnderlineNav-item hx_underlinenav-item no-wrap js-responsive-underlinenav-item\\" data-tab-item=\\"insights-tab\\" data-ga-click=\\"Repository, Navigation click, Insights tab\\" data-selected-links=\\"repo_graphs repo_contributors dependency_graph dependabot_updates pulse people /microsoft/playwright/pulse\\" href=\\"/microsoft/playwright/pulse\\">
              <svg classes=\\"UnderlineNav-octicon\\" display=\\"none inline\\" height=\\"16\\" class=\\"octicon octicon-graph UnderlineNav-octicon d-none d-sm-inline\\" viewBox=\\"0 0 16 16\\" version=\\"1.1\\" width=\\"16\\" aria-hidden=\\"true\\"><path fill-rule=\\"evenodd\\" d=\\"M1.5 1.75a.75.75 0 00-1.5 0v12.5c0 .414.336.75.75.75h14.5a.75.75 0 000-1.5H1.5V1.75zm14.28 2.53a.75.75 0 00-1.06-1.06L10 7.94 7.53 5.47a.75.75 0 00-1.06 0L3.22 8.72a.75.75 0 001.06 1.06L7 7.06l2.47 2.47a.75.75 0 001.06 0l5.25-5.25z\\"></path></svg>
            <span data-content=\\"Insights\\">Insights</span>
              <span title=\\"Not available\\" class=\\"Counter \\"></span>
</a>      </li>

</ul>        <div class=\\"position-absolute right-0 pr-3 pr-md-4 pr-lg-5 js-responsive-underlinenav-overflow\\" style=\\"visibility:hidden;\\">
      <details class=\\"details-overlay details-reset position-relative\\">
  <summary role=\\"button\\">
    <div class=\\"UnderlineNav-item mr-0 border-0\\">
            <svg class=\\"octicon octicon-kebab-horizontal\\" viewBox=\\"0 0 16 16\\" version=\\"1.1\\" width=\\"16\\" height=\\"16\\" aria-hidden=\\"true\\"><path d=\\"M8 9a1.5 1.5 0 100-3 1.5 1.5 0 000 3zM1.5 9a1.5 1.5 0 100-3 1.5 1.5 0 000 3zm13 0a1.5 1.5 0 100-3 1.5 1.5 0 000 3z\\"></path></svg>
            <span class=\\"sr-only\\">More</span>
          </div>
</summary>  <div>
    <details-menu role=\\"menu\\" class=\\"dropdown-menu dropdown-menu-sw \\">
  
            <ul>
                <li data-menu-item=\\"code-tab\\" hidden=\\"\\">
                  <a role=\\"menuitem\\" class=\\"js-selected-navigation-item selected dropdown-item\\" aria-current=\\"page\\" data-selected-links=\\" /microsoft/playwright\\" href=\\"/microsoft/playwright\\">
                    Code
</a>                </li>
                <li data-menu-item=\\"issues-tab\\" hidden=\\"\\">
                  <a role=\\"menuitem\\" class=\\"js-selected-navigation-item dropdown-item\\" data-selected-links=\\" /microsoft/playwright/issues\\" href=\\"/microsoft/playwright/issues\\">
                    Issues
</a>                </li>
                <li data-menu-item=\\"pull-requests-tab\\" hidden=\\"\\">
                  <a role=\\"menuitem\\" class=\\"js-selected-navigation-item dropdown-item\\" data-selected-links=\\" /microsoft/playwright/pulls\\" href=\\"/microsoft/playwright/pulls\\">
                    Pull requests
</a>                </li>
                <li data-menu-item=\\"actions-tab\\" hidden=\\"\\">
                  <a role=\\"menuitem\\" class=\\"js-selected-navigation-item dropdown-item\\" data-selected-links=\\" /microsoft/playwright/actions\\" href=\\"/microsoft/playwright/actions\\">
                    Actions
</a>                </li>
                <li data-menu-item=\\"projects-tab\\" hidden=\\"\\">
                  <a role=\\"menuitem\\" class=\\"js-selected-navigation-item dropdown-item\\" data-selected-links=\\" /microsoft/playwright/projects\\" href=\\"/microsoft/playwright/projects\\">
                    Projects
</a>                </li>
                <li data-menu-item=\\"security-tab\\" hidden=\\"\\">
                  <a role=\\"menuitem\\" class=\\"js-selected-navigation-item dropdown-item\\" data-selected-links=\\" /microsoft/playwright/security\\" href=\\"/microsoft/playwright/security\\">
                    Security
</a>                </li>
                <li data-menu-item=\\"insights-tab\\" hidden=\\"\\">
                  <a role=\\"menuitem\\" class=\\"js-selected-navigation-item dropdown-item\\" data-selected-links=\\" /microsoft/playwright/pulse\\" href=\\"/microsoft/playwright/pulse\\">
                    Insights
</a>                </li>
            </ul>

</details-menu>
</div></details>    </div>

"
`;
